backend/src/stwo/prover.rs

```diff
@@ -334,9 +401,37 @@ where
         tree_builder.commit(prover_channel);
 
         let mut tree_builder = commitment_scheme.tree_builder();
-        tree_builder.extend_evals(witness_by_machine);
+        tree_builder.extend_evals(witness_cols_circle_domain_eval.into_iter().flatten());
         tree_builder.commit(prover_channel);
 
+        let tree_span_provider = &mut TraceLocationAllocator::default();
+        // Each column size in machines needs its own component, the components from different machines are stored in this vector
+        let mut components = Vec::new();
+
+        // Build the circuit. The circuit includes constraints of all the machines in both stage 0 and stage 1
+        let mut constant_cols_offset_acc = 0;
+        self.split.iter().zip_eq(machine_log_sizes.iter()).for_each(
```

Comment:
```suggestion
        let components = self.split.iter().zip_eq(machine_log_sizes.iter()).map(
```

I think this functional style of writing code has a few advantages. For example, written like this it is obvious that there will be a component for each machine (you don't have to track the `push` calls), and also that they are independent of each other (computing one machine's component doesn't change a previous machine's component).

Reply:
With this, because `components` is no longer `mut`, the `components.iter_mut()` call won't work, but there is no reason not to use `iter()` AFAICS. Also, the `let components_slice = components_slice.as_mut_slice();` is not needed I think, you could just apss `&components_slice` to `prove`.

Reply:
nice! code updated.